Rate limit logspam during docker service pulls

dperny opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/docker/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**\- What I did**

During image pulls on docker service create, logs will only show status
updates and updates every 1 second on long-running actions like
downloading and extracting. Adds golang.org/x/time/rate as dependency.

Ports docker/swarmkit#1352 to docker/docker.

**\- How I did it**
Add check for status change and used rate to limit updates if there was no status change

**\- How to verify it**
With `-log-level debug` create a service with an image that must be pulled. The log should look like this:

```
DEBU[0055] Pulling ref from V2 registry: ubuntu:latest
DEBU[0055] pulling blob sha256:d8909ae88469beef31d7ad33c85a0a1b2ade2cef4d51bab0c1353abd8477054d
DEBU[0055] pulling blob sha256:2f0243478e1f7fcb578778370ebcbcb350ef637e8b966c6a24830df61a3e89b8
DEBU[0055] pull in progress
DEBU[0055] pull in progress                              image=ubuntu:latest status=Pulling fs layer
DEBU[0055] pulling blob sha256:820f09abed29b404920f73274dd7107a4080d22a40351683aa9c410df3d18eeb
DEBU[0055] pull in progress                              image=ubuntu:latest status=Waiting
DEBU[0055] pull in progress                              current=823 image=ubuntu:latest status=Downloading total=823
DEBU[0055] Downloaded d8909ae88469 to tempfile /var/lib/docker/tmp/GetImageBlob656252957
DEBU[0055] pulling blob sha256:01193a8f3d889f5c75dd4c54e7ad2a81216f10e7b2d515af571d17a4015b3d59
DEBU[0055] pull in progress                              image=ubuntu:latest status=Verifying Checksum
DEBU[0055] pull in progress                              image=ubuntu:latest status=Download complete
DEBU[0055] Downloaded 820f09abed29 to tempfile /var/lib/docker/tmp/GetImageBlob025140823
DEBU[0055] pull in progress                              current=501200 image=ubuntu:latest status=Downloading total=49678667
DEBU[0056] pull in progress                              current=4056528 image=ubuntu:latest status=Downloading total=49678667
DEBU[0056] Downloaded 01193a8f3d88 to tempfile /var/lib/docker/tmp/GetImageBlob884477898
DEBU[0056] pull in progress                              image=ubuntu:latest status=Verifying Checksum
DEBU[0056] pull in progress                              image=ubuntu:latest status=Download complete
DEBU[0056] pull in progress                              current=8119760 image=ubuntu:latest status=Downloading total=49678667
DEBU[0057] pull in progress                              current=21833168 image=ubuntu:latest status=Downloading total=49678667
DEBU[0058] pull in progress                              current=44180944 image=ubuntu:latest status=Downloading total=49678667
DEBU[0058] Downloaded 2f0243478e1f to tempfile /var/lib/docker/tmp/GetImageBlob218862296
DEBU[0058] pull in progress                              image=ubuntu:latest status=Verifying Checksum
DEBU[0058] pull in progress                              current=524288 image=ubuntu:latest status=Extracting total=49678667
DEBU[0059] pull in progress                              current=13107200 image=ubuntu:latest status=Extracting total=49678667
DEBU[0060] pull in progress                              current=30408704 image=ubuntu:latest status=Extracting total=49678667
DEBU[0061] pull in progress                              current=45613056 image=ubuntu:latest status=Extracting total=49678667
DEBU[0061] Applied tar sha256:d8d865b23727fa63dc7d4e9f02efc3e2433c8f220972689c6b6ba09e8136c162 to 232fadc8c56285c99fc4b2a7818ed5cbf169e29ba0948d8b58840b3bafd986aa, size: 126391956
DEBU[0061] pull in progress                              image=ubuntu:latest status=Pull complete
DEBU[0061] pull in progress                              current=823 image=ubuntu:latest status=Extracting total=823
DEBU[0061] Applied tar sha256:3834bde7e567c6fa5bb092531b4f769ad97a2d0b46d9ffb8a79922206d9db560 to 5eb4913a4bfc06df815cf0b78c8ffa5c5f86bc2505853b02eb72043b6c03075d, size: 745
DEBU[0061] pull in progress                              image=ubuntu:latest status=Pull complete
DEBU[0061] pull in progress                              current=445 image=ubuntu:latest status=Extracting total=445
DEBU[0062] Applied tar sha256:2bed5b3ec49f8e13e4c746fca9dd0807a5ae5fa37f3b8da7455a6aa1507706b7 to 79bec3e76986b2d11bfc5018b9863844c32a32f0ea30dbfa20e031ded5f98967, size: 0
DEBU[0062] pull in progress                              image=ubuntu:latest status=Pull complete
DEBU[0062] pull in progress                              current=682 image=ubuntu:latest status=Extracting total=682
DEBU[0062] Applied tar sha256:4699cbd1a947d501030bdfa40ec4a23c706e6ffc20922c74e8d5d8352dfdaac3 to f9a8971c39b0b1d21ad21eb67c759d77b2e9398c0fba8b6ea0d29c03004d93a0, size: 1895
DEBU[0062] pull in progress                              image=ubuntu:latest status=Pull complete
DEBU[0062] pull in progress                              image=ubuntu:latest status=Digest: sha256:8e2324f2288c26e1393b63e680ee7844202391414dbd48497e9a4fd997cd3cbf
DEBU[0062] pull in progress                              image=ubuntu:latest status=Status: Downloaded newer image for ubuntu:latest
DEBU[0062] container mounted via layerStore: /var/lib/docker/aufs/mnt/c618c83d182f8a9cdaf5b5fd0dc7567797ae8ab2b52ec6241bd8fb6fc4102b7c
DEBU[0062] state changed                                 module=node/agent/taskmanager state.desired=RUNNING state.transition=PREPARING->READY task.id=581ag6qxi9pikkoz4ehe13l75
```

**\- Description for the changelog**

<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

Made Swarm mode service image pulls spam debug logs less

**\- A picture of a cute animal (not mandatory but encouraged)**
![bird](https://cloud.githubusercontent.com/assets/2367858/17605974/e0942ede-5fd1-11e6-93b1-4190e526dc82.jpg)

dperny committed:
Rate limit logspam during docker service pulls

During image pulls on docker service create, logs will only show status
updates and updates every 1 second on long-running actions like
downloading and extracting. Adds golang.org/x/time/rate as dependency.

Ports docker/swarmkit#1352 to docker/docker.

Signed-off-by: Drew Erny <drew.erny@docker.com>

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "rate-limit-swarm-pull-log" git@github.com:dperny/docker.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "rate-limit-swarm-pull-log" git@github.com:dperny/docker.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

dperny commented:
PTAL closely, I have updated vendoring and am not totally confident that it was done correctly.

Also I added the signoff but Gordon thinks I haven't.

aaronlehmann commented on daemon/cluster/executor/container/adapter.go:
```diff
@@ -65,17 +66,44 @@ func (c *containerAdapter) pullImage(ctx context.Context) error {
 	dec := json.NewDecoder(pr)
+	dec.UseNumber()
 	m := map[string]interface{}{}
+	spamLimiter := rate.NewLimiter(rate.Every(1000*time.Millisecond), 1)
```
nit: `time.Second`

dperny replied:
You're right.

thaJeztah replied:
in stead of rate limiting, couldn't we simply print every xx percent (10% 20%)?

dperny replied:
@thaJeztah IMHO not a substantial enough improvement for me to spend an 30minutes - hour reworking it for percents instead of rates, but if you think it's important I am willing to do so

thaJeztah replied:
@dperny I'm fine either way I think, but saw the extra dependencies for implementing the rate-limit, so thought that print every xx percent would possibly be a simpler solution.

aaronlehmann commented:
LGTM if CI passes

dperny commented:
`1000*time.Millisecond` -> `time.Second`

dperny commented:
it's failing `vendor`, how can I test and fix this?

dperny commented:
In my local environment, doing `make shell` and then `hack/make.sh validate-vendor` returns `Congratulations! All vendoring changes are done the right way.`

aaronlehmann commented:
Just looks like a network glitch. I've restarted that build.

cpuguy83 commented on daemon/cluster/executor/container/adapter.go:
```diff
@@ -65,17 +66,44 @@ func (c *containerAdapter) pullImage(ctx context.Context) error {
 	}()
 
 	dec := json.NewDecoder(pr)
+	dec.UseNumber()
```
How come this?

aaronlehmann replied:
Without `UseNumber`, the progress information gets unmarshalled to floats, and gets printed in scientific notation by logrus.

https://github.com/docker/swarmkit/pull/1352#discussion_r74291601

cpuguy83 commented:
LGTM
